# Real data integration

Here we fill the database with real data instead of only our mock data.

We used two sources:
- **A: Kamernet listings** (Jul 2019 - Mar 2020, from Kaggle). Real rooms, studios and apartments with rent, size, type and postal code. Used for `Address`, `Accommodation` and the rent numbers in `City`.
- **B: DUO enrolments** (WO + HBO). How many students are enrolled per city. Used for `City.student_population`.

A shows what housing is available and B shows how many students need housing, so the two complement each other and neither one contains the other.

The raw files are cleaned by `scripts/map_real_data.py`, which writes the csv files in `data/clean/`. This notebook loads those csv files into MySQL.

**Run `housing_database.ipynb` first**, it creates the tables and the mock data.

Section 7 adds **AI-generated mock data** for the tables our real sources don't cover. It is clearly marked as AI.

## 1. Connecting to the database

In [1]:
import pandas as pd
from sqlalchemy import create_engine, text

# change this to your own mysql username and password
engine = create_engine("mysql+pymysql://root:12345@localhost/housing_database")

%load_ext sql
%sql mysql+pymysql://root:12345@localhost/housing_database

## 2. Reading the cleaned csv files

In [2]:
city = pd.read_csv("data/clean/city.csv")
address = pd.read_csv("data/clean/address.csv")
accommodation = pd.read_csv("data/clean/accommodation.csv")

print(len(city), "cities,", len(address), "addresses,", len(accommodation), "accommodations")
city

10 cities, 4649 addresses, 7304 accommodations


,city_code,euro_sqm,min_rent,max_rent,student_population
0,AMSTERDAM,42.01,100.0,5000.0,130227
1,DEN HAAG,27.54,200.0,5000.0,33304
2,EINDHOVEN,23.21,174.0,2250.0,36055
3,GRONINGEN,23.15,100.0,1950.0,59559
4,LEIDEN,26.23,200.0,2950.0,38261
5,MAASTRICHT,21.81,225.0,1850.0,27225
6,NIJMEGEN,24.64,225.0,2250.0,46712
7,ROTTERDAM,28.98,100.0,2300.0,73427
8,TILBURG,22.02,155.0,2275.0,30880
9,UTRECHT,28.31,100.0,2800.0,73620


## 3. Making the data fit our schema

The csv columns are named a bit differently than our table, so we rename them.

In Kamernet `furnished` is text (Furnished / Unfurnished / Uncarpeted) but in our table it is a BOOL, so only "Furnished" becomes true. Uncarpeted means the place is completely empty, so that counts as not furnished.

The deposit and the Kamernet id don't have a column in Accommodation (deposit belongs to a Contract), so we leave them out.

In [3]:
accommodation = accommodation.rename(columns={
    "accommodation_type": "accomodation_type",
    "furnished": "accomodation_furnished",
})
accommodation["accomodation_furnished"] = accommodation["accomodation_furnished"].map(
    {"Furnished": True, "Unfurnished": False, "Uncarpeted": False}
)
accommodation = accommodation.drop(columns=["deposit", "source_id"])

accommodation.head()

,accommodation_id,landlord_id,agency_id,address_id,subsidy_possible,square_meters,rent_price,accomodation_type,accomodation_furnished
0,1001,NaN,NaN,1001,no,14,500,room,False
1,1002,NaN,NaN,1002,no,30,950,studio,True
2,1003,NaN,NaN,1003,no,11,1000,room,True
3,1004,NaN,NaN,1004,no,22,475,room,False
4,1005,NaN,NaN,1005,no,14,340,room,False


## 4. Inserting the data

Everything is done in one transaction, so if one insert fails nothing ends up half inserted.

The real data starts at id 1001 so it doesn't clash with the mock data (ids 1 to 60). We first delete real data from an earlier run, so this cell can be run more than once.

All 10 cities already exist from the mock data, so for those we update the numbers instead of inserting them again.

In [4]:
with engine.begin() as con:
    # remove real data from an earlier run, the mock data stays
    # (contracts on real listings come from section 7 and have to go first because of the foreign key)
    con.execute(text("DELETE FROM Contract WHERE accommodation_id > 1000"))
    con.execute(text("DELETE FROM Accommodation WHERE accommodation_id > 1000"))
    con.execute(text("DELETE FROM Address WHERE address_id > 1000"))

    # cities: insert new ones, update the ones we already have
    con.execute(text("""
        INSERT INTO City (city_code, euro_sqm, min_rent, max_rent, student_population)
        VALUES (:city_code, :euro_sqm, :min_rent, :max_rent, :student_population) AS new
        ON DUPLICATE KEY UPDATE
            euro_sqm = new.euro_sqm,
            min_rent = new.min_rent,
            max_rent = new.max_rent,
            student_population = new.student_population
    """), city.to_dict("records"))

    address.to_sql("Address", con, if_exists="append", index=False)
    accommodation.to_sql("Accommodation", con, if_exists="append", index=False)

print("done")

done


## 5. Checking the result

In [5]:
%%sql
SELECT 'City' AS table_name, COUNT(*) AS total_rows FROM City
UNION ALL SELECT 'Address', COUNT(*) FROM Address
UNION ALL SELECT 'Accommodation', COUNT(*) FROM Accommodation;

 * mysql+pymysql://root:***@localhost/housing_database
3 rows affected.


table_name,total_rows
City,10
Address,4709
Accommodation,7364


In [6]:
%%sql
SELECT * FROM City;

 * mysql+pymysql://root:***@localhost/housing_database
10 rows affected.


city_code,euro_sqm,min_rent,max_rent,student_population
AMSTERDAM,42.01,100.00,5000.00,130227
DEN HAAG,27.54,200.00,5000.00,33304
EINDHOVEN,23.21,174.00,2250.00,36055
GRONINGEN,23.15,100.00,1950.00,59559
LEIDEN,26.23,200.00,2950.00,38261
MAASTRICHT,21.81,225.00,1850.00,27225
NIJMEGEN,24.64,225.00,2250.00,46712
ROTTERDAM,28.98,100.00,2300.00,73427
TILBURG,22.02,155.00,2275.00,30880
UTRECHT,28.31,100.00,2800.00,73620


## 6. Queries on the real data

### How many students are there per listing in each city?
This one uses both sources: the student numbers from DUO and the listings from Kamernet.

In [7]:
%%sql
SELECT c.city_code,
       c.student_population,
       COUNT(a.accommodation_id) AS listings,
       ROUND(c.student_population / COUNT(a.accommodation_id)) AS students_per_listing,
       ROUND(AVG(a.rent_price)) AS avg_rent
FROM City c
JOIN Address ad ON ad.city_code = c.city_code
JOIN Accommodation a ON a.address_id = ad.address_id
WHERE a.accommodation_id > 1000 -- only real data
GROUP BY c.city_code, c.student_population
ORDER BY students_per_listing DESC;

 * mysql+pymysql://root:***@localhost/housing_database
10 rows affected.


city_code,student_population,listings,students_per_listing,avg_rent
NIJMEGEN,46712,273,171,575
LEIDEN,38261,299,128,555
UTRECHT,73620,602,122,745
EINDHOVEN,36055,401,90,709
MAASTRICHT,27225,327,83,519
AMSTERDAM,130227,1988,66,976
TILBURG,30880,482,64,534
DEN HAAG,33304,545,61,835
ROTTERDAM,73427,1298,57,693
GRONINGEN,59559,1089,55,522


### Average rent and size per accommodation type

In [8]:
%%sql
SELECT accomodation_type,
       COUNT(*) AS listings,
       ROUND(AVG(rent_price)) AS avg_rent,
       ROUND(AVG(square_meters)) AS avg_sqm,
       ROUND(AVG(rent_price / square_meters), 2) AS avg_euro_per_sqm
FROM Accommodation
WHERE accommodation_id > 1000
GROUP BY accomodation_type
ORDER BY avg_rent DESC;

 * mysql+pymysql://root:***@localhost/housing_database
4 rows affected.


accomodation_type,listings,avg_rent,avg_sqm,avg_euro_per_sqm
apartment,1956,1251,68,19.97
studio,605,759,30,27.08
room,4738,515,17,34.73
anti-kraak,5,454,38,11.77


### How many real listings fit each of our students?
Same idea as the matching query in the main notebook, but now with the real listings: in the city the student is searching in, within their budget and of a type they want.

In [9]:
%%sql
SELECT s.student_id, s.search_city_code, s.max_rent,
       COUNT(DISTINCT a.accommodation_id) AS matching_listings
FROM Student s
JOIN StudentPreference p ON p.student_id = s.student_id
LEFT JOIN Address ad ON ad.city_code = s.search_city_code
LEFT JOIN Accommodation a ON a.address_id = ad.address_id
    AND a.accomodation_type = p.accommodation_type
    AND a.rent_price <= s.max_rent
    AND a.accommodation_id > 1000
WHERE s.student_id < 90001 -- only our own mock students, the AI ones are in section 7
GROUP BY s.student_id, s.search_city_code, s.max_rent
ORDER BY matching_listings DESC;

 * mysql+pymysql://root:***@localhost/housing_database


9 rows affected.


student_id,search_city_code,max_rent,matching_listings
4,ROTTERDAM,850.00,844
8,GRONINGEN,700.00,750
1,MAASTRICHT,600.00,230
5,UTRECHT,500.00,198
3,AMSTERDAM,900.00,54
7,EINDHOVEN,1000.00,39
2,MAASTRICHT,750.00,26
6,DEN HAAG,650.00,12
9,NIJMEGEN,450.00,2


## 7. AI generated Script to create Mock Data that fits our realworld DATA 


In [ ]:
import numpy as np

rng = np.random.default_rng(42)  # fixed seed, so every run makes the same data
AI_ID = 90001  # all AI rows get ids from here on

# real listings with their city, used to make the AI data fit
listings = pd.read_csv("data/clean/accommodation.csv").merge(address[["address_id", "city_code"]], on="address_id")

# Agency: 2 per city
agency_city = [c for c in city.city_code for _ in range(2)]
agency_ids = range(AI_ID, AI_ID + len(agency_city))
agency = pd.DataFrame({
    "agency_id": agency_ids,
    "address_id": None,
    "name": [f"{c.title()} Student Rooms {n % 2 + 1} (AI mock)" for n, c in enumerate(agency_city)],
    "phone_number": [f"+31003{i:05d}" for i in agency_ids],
    "email": [f"agency{i}@ai-mock.example" for i in agency_ids],
})

# Landlord: 70% work with an agency, the rest rent out on their own
first_names = ["Jan", "Anna", "Pieter", "Sanne", "Mohamed", "Lisa", "Daan", "Emma", "Ahmet", "Femke", "Thomas", "Sophie"]
last_names = ["de Jong", "Jansen", "de Vries", "van Dijk", "Bakker", "Visser", "Smit", "Meijer", "Bos", "Mulder", "Yilmaz", "Peters"]
landlord_ids = range(AI_ID, AI_ID + 80)
landlord = pd.DataFrame({
    "landlord_id": landlord_ids,
    "agency_id": [int(rng.choice(agency_ids)) if rng.random() < 0.7 else None for _ in landlord_ids],
    "address_id": None,
    "name": [f"{rng.choice(first_names)} {rng.choice(last_names)} (AI mock)" for _ in landlord_ids],
    "phone_number": [f"+31002{i:05d}" for i in landlord_ids],
    "email": [f"landlord{i}@ai-mock.example" for i in landlord_ids],
})

# Student: spread over the cities like the real DUO numbers
n_students = 500
student_ids = range(AI_ID, AI_ID + n_students)
student_city = rng.choice(city.city_code.to_numpy(), size=n_students,
                          p=(city.student_population / city.student_population.sum()).to_numpy())

# budget = a real room/studio rent in their city, give or take a bit, rounded to 25
cheap = listings[listings.accommodation_type.isin(["room", "studio"])]
budget = [round(rng.choice(cheap.rent_price[cheap.city_code == c].to_numpy()) * rng.uniform(0.9, 1.4) / 25) * 25
          for c in student_city]

nationalities = ["Dutch", "German", "Belgian", "Italian", "Spanish", "French", "Polish", "Romanian", "Chinese", "Indian", None]
student = pd.DataFrame({
    "student_id": student_ids,
    "nationality": rng.choice(np.array(nationalities, dtype=object), size=n_students,
                              p=[.55, .10, .04, .05, .04, .04, .04, .03, .04, .04, .03]),
    "phone_number": [f"+31001{i:05d}" for i in student_ids],
    "email": [f"student{i}@ai-mock.example" for i in student_ids],
    "max_rent": budget,
    "search_city_code": student_city,
})

# StudentPreference: every student wants 1 or 2 types, rooms and studios are the most popular
types = ["room", "studio", "apartment", "anti-kraak"]
preference = pd.DataFrame(
    [(s, str(t)) for s in student_ids
     for t in rng.choice(types, size=rng.integers(1, 3), replace=False, p=[.5, .3, .15, .05])],
    columns=["student_id", "accommodation_type"],
)

# Contract: about 60% of the students found a real listing that fits them
free = listings.set_index("accommodation_id")  # every listing can only be rented once
contracts = []
for s in student.itertuples():
    if rng.random() > 0.6:
        continue
    wanted = preference.accommodation_type[preference.student_id == s.student_id]
    options = free[(free.city_code == s.search_city_code) & (free.rent_price <= s.max_rent)
                   & free.accommodation_type.isin(wanted)]
    if options.empty:
        continue
    acc_id = int(rng.choice(options.index))
    start = pd.Timestamp("2019-09-01") + pd.DateOffset(months=int(rng.integers(0, 7)))
    min_duration = int(rng.choice([3, 6, 12]))
    max_duration = min_duration * int(rng.choice([1, 2]))
    end = start + pd.DateOffset(months=max_duration) - pd.Timedelta(days=1)
    contracts.append({
        "contract_id": AI_ID + len(contracts),
        "student_id": s.student_id,
        "accommodation_id": acc_id,
        "start_date": start.date(),
        "end_date": end.date(),
        "min_duration": min_duration,
        "max_duration": max_duration,
        "cancellation_time": (end - pd.DateOffset(months=1)).date(),
        "deposit": free.deposit[acc_id],  # real deposit from Kamernet
    })
    free = free.drop(acc_id)
contract = pd.DataFrame(contracts)



20 agencies, 80 landlords, 500 students, 727 preferences, 271 contracts (all AI generated)


,student_id,nationality,phone_number,email,max_rent,search_city_code
0,90001,NaN,+3100190001,student90001@ai-mock.example,925,DEN HAAG
1,90002,Dutch,+3100190002,student90002@ai-mock.example,800,UTRECHT
2,90003,Dutch,+3100190003,student90003@ai-mock.example,1175,DEN HAAG
3,90004,Dutch,+3100190004,student90004@ai-mock.example,750,AMSTERDAM
4,90005,Dutch,+3100190005,student90005@ai-mock.example,675,TILBURG


In [ ]:
with engine.begin() as con:
   
    agency.to_sql("Agency", con, if_exists="append", index=False)
    landlord.to_sql("Landlord", con, if_exists="append", index=False)
    student.to_sql("Student", con, if_exists="append", index=False)
    preference.to_sql("StudentPreference", con, if_exists="append", index=False)
    contract.to_sql("Contract", con, if_exists="append", index=False)


AI mock data inserted


### Mock rows vs AI rows per table

In [12]:
%%sql
SELECT 'Agency' AS table_name, SUM(agency_id < 90001) AS mock_rows, SUM(agency_id >= 90001) AS ai_rows FROM Agency
UNION ALL SELECT 'Landlord', SUM(landlord_id < 90001), SUM(landlord_id >= 90001) FROM Landlord
UNION ALL SELECT 'Student', SUM(student_id < 90001), SUM(student_id >= 90001) FROM Student
UNION ALL SELECT 'StudentPreference', SUM(student_id < 90001), SUM(student_id >= 90001) FROM StudentPreference
UNION ALL SELECT 'Contract', SUM(contract_id < 90001), SUM(contract_id >= 90001) FROM Contract;

 * mysql+pymysql://root:***@localhost/housing_database
5 rows affected.


table_name,mock_rows,ai_rows
Agency,10,20
Landlord,10,80
Student,13,500
StudentPreference,10,727
Contract,10,271
